akka-actor/src/main/scala/akka/routing/DefaultMetricsBasedResizer.scala

```diff
@@ -0,0 +1,222 @@
+      downsizeAfterUnderutilizedFor = resizerCfg.getDuration("downsize-after-underutilized-for"),
+      exploreStepSize = resizerCfg.getDouble("explore-step-size"),
+      explorationProbability = resizerCfg.getDouble("chance-of-exploration"),
+      weightOfLatestMetric = resizerCfg.getDouble("weight-of-latest-metric"),
+      downsizeRatio = resizerCfg.getDouble("downsize-ratio"))
+
+}
+
+/**
+ * Implementation of [[Resizer]] that adjust the [[Pool]]
+ * based on performance per size.
+ * It keeps a performance log so it's stateful as well as
+ * a larger memory footprint than the [[DefaultResizer]].
+ * For documentation about the the parameters, see the reference.conf -
+ * akka.actor.deployment.default.metrics-based-resizer
+ *
+ * CAUTION: [[resize]] also changes the inner state of the resizer,
+ * do not call it unnecessarily or even concurrently
+ *
+ *
+ */
+@SerialVersionUID(1L)
+case class DefaultMetricsBasedResizer(
+  lowerBound: PoolSize = 1,
+  upperBound: PoolSize = 30,
+  chanceOfScalingDownWhenFull: Double = 0.2,
+  actionInterval: Duration = Duration.ofSeconds(5),
+  numOfAdjacentSizesToConsiderDuringOptimization: Int = 6,
+  exploreStepSize: Double = 0.1,
+  downsizeRatio: Double = 0.8,
+  downsizeAfterUnderutilizedFor: Duration = Duration.ofHours(72),
+  explorationProbability: Double = 0.4,
+  weightOfLatestMetric: Double = 0.5) extends MetricsBasedResizer {
+  /**
+   * Leave package accessible for testing purpose
+   */
+  private[routing] var performanceLog: PerformanceLog = Map.empty
+  /**
+   * Leave package accessible for testing purpose
+   */
+  private[routing] var record: ResizeRecord = ResizeRecord()
+
+  private val random = ThreadLocalRandom.current()
+
+  private def checkParamAsProbability(value: Double, paramName: String): Unit =
+    if (value < 0 || value > 1) throw new IllegalArgumentException(s"$paramName must be between 0 and 1 (inclusive), was: [%s]".format(value))
+
+  private def checkParamAsPositiveNum(value: Double, paramName: String): Unit =
+    if (value < 0) throw new IllegalArgumentException(s"$paramName must be >= 0, was: [%s]".format(value))
+
+  checkParamAsPositiveNum(lowerBound, "lowerBound")
+  checkParamAsPositiveNum(upperBound, "upperBound")
+  if (upperBound < lowerBound) throw new IllegalArgumentException("upperBound must be >= lowerBound, was: [%s] < [%s]".format(upperBound, lowerBound))
+
+  checkParamAsProbability(chanceOfScalingDownWhenFull, "chanceOfScalingDownWhenFull")
+  checkParamAsPositiveNum(numOfAdjacentSizesToConsiderDuringOptimization, "numOfAdjacentSizesToConsiderDuringOptimization")
+  checkParamAsPositiveNum(exploreStepSize, "exploreStepSize")
+  checkParamAsPositiveNum(downsizeRatio, "downsizeRatio")
+  checkParamAsProbability(explorationProbability, "explorationProbability")
+  checkParamAsProbability(weightOfLatestMetric, "weightOfLatestMetric")
+
+  //accessible only for testing purpose
+
+  val historySampleRate: Duration = actionInterval dividedBy 100
```

Comment:
Actually this one should probably better be configurable. Will make the change once I get back to my laptop.